system/mp/mpcomm.cpp

```diff
@@ -441,97 +444,322 @@ class CMPServer;
+        id[0].set(hostEp);
+        id[1].set(remoteEp);
+
+        hdr.size = sizeof(PacketHeader);
+        hdr.tag = TAG_SYS_BCAST;
+        hdr.flags = 0;
+        hdr.version = MP_PROTOCOL_VERSION;
+        setRole(role);
+    }
+    ConnectHdr()
+    {
+    }
+    SocketEndpointV4 id[2];
+    PacketHeader hdr;
+    inline void setRole(unsigned __int64 role)
+    {
+        hdr.replytag = (mptag_t) (role >> 32);
+        hdr.sequence = (unsigned) (role & 0xffffffff);
+    }
+    inline unsigned __int64 getRole() const
+    {
+        return (((unsigned __int64)hdr.replytag)<<32) | ((unsigned __int64)hdr.sequence);
+    }
+};
 
-    Owned<ISecureSocketContext> secureContextServer;
 
-    class CSlowClientProcessor : implements IThreaded
+class CMPConnectThread: public Thread
+{
+    class CConnectSelectHandler
     {
         CMPConnectThread &owner;
-        CThreaded threaded;
-        std::vector<Owned<ISocket>> slowClientsSocks;
-        CriticalSection crit;
-        Semaphore sem;
-        std::atomic<bool> stopped = true;
-
+        Owned<ISocketSelectHandler> selectHandler;
+        unsigned mode = SELECTMODE_READ;
     public:
-        CSlowClientProcessor(CMPConnectThread &_owner) : threaded("CSlowClientProcessor"), owner(_owner)
+        class CSocketHandler : public CInterfaceOf<ISocketSelectNotify>
         {
-        }
-        void start()
+            CConnectSelectHandler &selectHandler;
+            Owned<ISocket> sock;
+            SocketEndpoint peerEP;
+            mutable StringBuffer peerHostText, peerEndpointText;
+            ConnectHdr hdr;
+            cycle_t createTime = 0;
+            size32_t readSoFar = 0;
+            CriticalSection crit;
+            bool closedOrHandled = false;
+        public:
+            CSocketHandler(CConnectSelectHandler &_selectHandler, ISocket *_sock, const SocketEndpoint &_peerEP) : selectHandler(_selectHandler), sock(_sock), peerEP(_peerEP)
+            {
+                createTime = get_cycles_now();
+            }
+            ISocket *querySocket()
+            {
+                return sock;
+            }
+            ConnectHdr &queryHdr()
+            {
+                return hdr;
+            }
+            cycle_t queryCreateTime() const
+            {
+                return createTime;
+            }
+            size32_t queryReadSoFar() const
+            {
+                return readSoFar;
+            }
+            const char *queryPeerHostText() const
+            {
+                if (0 == peerHostText.length())
+                    peerEP.getHostText(peerHostText);
+                return peerHostText;
+            }
+            const char *queryPeerEndpointText() const
+            {
+                if (0 == peerEndpointText.length())
+                    peerEP.getEndpointHostText(peerEndpointText);
+                return peerEndpointText;
+            }
+            bool closeIfTimedout(cycle_t now)
+            {
+                return false;
```

Comment:
>If a legacy client tries to connect to Dali (ie sending only 12 bytes) or perhaps some rogue client - how long would Dali/MP keep the connection open waiting for the remaining data ? In my tests the connection seems to remain open for a very long time.

@mckellyln - that would explain why it isn't closing them! :) [ I must have left in whilst testing at some point ]
Will remove!